sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,498 @@
+images is recorded instead.
+
+After device code split, all non-`SYCL_EXTERNAL` functions and copied
+`SYCL_EXTERNAL` functions left in device images with kernels marked with
+internal linkage to avoid multiple definition errors during runtime linking.
+After that `sycl-post-link` records list of names of exported functions, i.e.
+functions with `sycl-module-id` attribute and external linkage.
+
+In order to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its
+name, except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+    since they represent SPIR-V operations and will be transformed to SPIR-V
+    instructions during LLVM->SPIR-V translation.
+- Based on some attributes (which could be defined later) we may want to avoid
+  listing some functions as imported ones
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on separated images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represented by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
```

Comment:
I suggest that we add a sub-section about this `PI API` below: this API is important, as this is a part of interface between plugins and DPC++ RT and we don't even have its name listed here.

The sub-section could describe the API, its behavior (whether we assume that it is capable to link native binaries regardless of device or that it is some optional capability which should be checked before usage), possible implementation/limitations.

Reply:
Added, please see lines  236-259.